op-service/log/context_handler.go

```diff
@@ -0,0 +1,89 @@
+package log
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/ethereum-optimism/optimism/op-service/logmods"
+)
+
+type contextLogAttr struct {
+	name string
+	key  any
+}
+
+type logKeyIndexCtxType struct{}
+
+var contextLogAttrIndexCtxKey = logKeyIndexCtxType{}
+
+// LogKeyIndexFromContext reads a list of (name, key) pairs from the context
+// where the keys can be used to read other context values which must implement
+// slog.LogValuer and are used by contextHandler to add attributes to log records.
+func LogKeyIndexFromContext(ctx context.Context) []contextLogAttr {
+	v := ctx.Value(contextLogAttrIndexCtxKey)
+	if v == nil {
+		return nil
+	}
+	return v.([]contextLogAttr)
+}
+
+// RegisterLogAttrOnContext configures the context so that if a contextHandler
+// is in the log handler chain it will add an attr with key `name` and
+// value of `ctx.Value(key)` to all log records.
+func RegisterLogAttrOnContext(ctx context.Context, name string, key any) context.Context {
+	prevIndex := LogKeyIndexFromContext(ctx)
+	// prevIndex is possibly nil, we want to force copy it
+	nextIndex := append(append([]contextLogAttr{}, prevIndex...), contextLogAttr{name: name, key: key})
```

Comment:
Maybe worth commenting that we also need to copy to handle the forked context case (since child-contexts might otherwise append into the same unused parent slice capacity).

Reply:
> since child-contexts might otherwise append into the same unused parent slice capacity

This is negated by the fact the fact we are always appending to a completely new slice `[]contextLogAttr{}`, right?

Reply:
Yes, the outcome is correct, but the comment above "prevIndex is possibly nil, we want to force copy it" implies that it's just for the nil case, whereas this append mutation footgun is a more important reason to copy.